core/IQueue.go

```diff
@@ -0,0 +1,111 @@
+// Copyright (c) 2008-2018, Hazelcast, Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License")
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package core
+
+// IQueue is a concurrent, blocking, distributed, observable queue. Queue is not a partitioned data-structure.
+// All of the Queue content is stored in a single machine (and in the backup).
+// Queue will not scale by adding more members in the cluster.
+type IQueue interface {
+
+	// IDistributedObject is the base interface for all distributed objects.
+	IDistributedObject
+
+	// Add adds the given item to the end of the queue. Operation is successful only
+	// if queue has required capacity.
+	// Add returns true if the item is added, false otherwise.
```

Comment:
If the item cannot be added this method should set `err` with an appropriate error. If `err` is set, return value should not matter. If the item is added it returns `true`, otherwise `err` is set. Therefore this method never returns `false`.